perf(metrics-extraction): Cache should_use_on_demand decision

k-fish opened the pull request:
### Summary
Checking whether a query should use on-demand requires us to parse the query and use the event_search_grammar every time. This isn't that costly when doing an api call (in the 10ms to 100ms range) but when doing thousands of checks it adds up to a significant amount of time. This caches the decision of using on-demand solely based on the properties for up to an hour.

Behind an option so we can check the impact of this and roll it off quickly if required.

wmak reviewed (approved):
should hash the cache_key before using it, but otherwise lgtm

armenzg reviewed (approved):
💃🏻

k-fish committed:
perf(metrics-extraction): Cache should_use_on_demand decision

Checking whether a query should use on-demand requires us to parse the query and use the event_search_grammar every time. This isn't that costly when doing an api call (in the 10ms to 100ms range) but when doing thousands of checks it adds up to a significant amount of time. This caches the decision of using on-demand solely based on the properties for up to an hour.

k-fish committed:
Should be rate
